src/update-utmp/update-utmp.c

```diff
@@ -97,8 +97,15 @@ static int get_current_runlevel(Context *c) {
                                 "ActiveState",
                                 &error,
                                 &state);
+                if (sd_bus_error_has_names(&error,
+                                           SD_BUS_ERROR_NO_REPLY,
+                                           SD_BUS_ERROR_DISCONNECTED)) {
+                        /* systemd might have dropped off momentarily, let's not make this an error. */
+                        log_debug_errno(r, "Failed to get state of %s, ignoring: %s", table[i].special, bus_error_message(&error, r));
+                        continue;
```

Comment:
I don't think that's useful. We don't return an error, but we don't return a correct answer either.

Maybe it should do something like try up to 5 times with a short sleep inbetween?

A more general solution would be to keep the connection open at all times. But I guess we can't do this with dbus, because we need to destroy the state in the old binary and reconnect and recreate the state in the new binary. @poettering, comments?

Reply:
In the revised version, now it returns `0` on such case. That is handled as non-critical error by the caller. PTAL.

Reply:
Or, how about to fail the executable as is, but set `Restart=on-failure` to the service?